### ISSUE TITLE
Federate the Android and iOS implementations of 1P plugins

### ISSUE THREAD

**Issue by MisterJimson** [flutter/flutter#68498]
## Use case
From the flutter docs: https://flutter.dev/docs/development/packages-and-plugins/developing-packages#federated-plugins

```
Federated plugins are a a way of splitting support for different platforms into separate packages. So, a federated plugin can use one package for iOS, another for Android, another for web, and yet another for a car (as an example of an IoT device). Among other benefits, this approach allows a domain expert to extend an existing plugin to work for the platform they know best.
```

This seems at odds with the official plugins not having the Android and iOS implementations in separate projects.

Looking at `shared_preferences` as an example: https://github.com/flutter/plugins/tree/master/packages/shared_preferences

Why are web and desktop in separate packages but Android and iOS are not? If we want to swap the implementation of Android or iOS, the default implementation is still present. Why are they special?

## Proposal
Treat all platforms the same, 1 package per platform.

Here is an example of what I am calling "fully federated", where there is no platform implementations in the app facing package: https://github.com/MisterJimson/federated_plugins_example/tree/fully-federated

**Comment by stuartmorgan** [comment]
> Why are web and desktop in separate packages but Android and iOS are not?

Because they predate federation.

> Treat all platforms the same, 1 package per platform.

This is the plan of record (or at least it was; @amirh that's still the case, right?). However, the work of transitioning existing plugin implementations is low priority.

I expect that patches federating existing mobile implementations would be welcome.

**Comment by Sunbreak** [comment]
Does it work out? https://github.com/woodemi/quick_flutter.brick/issues/22#issuecomment-898800261

> - Choice B: all implementations are in their own federated package
>
> ```
> $ mason make quick_flutter_federated_plugin
> ...
> $ ls fedplugin/
> fedplugin			fedplugin_ios			fedplugin_macos			fedplugin_web
> fedplugin_android		fedplugin_linux			fedplugin_platform_interface	fedplugin_windows
> ```

**Comment by stuartmorgan** [comment]
I'm not sure what you are asking. If it's "is it possible to implement what this issue describes", then yes.

**Comment by stuartmorgan** [comment]
Note: `camera` is blocked by https://github.com/flutter/flutter/issues/100555.
